lib/ll_client.ex

```diff
@@ -0,0 +1,173 @@
+defmodule Jellygrinder.LLClient do
+  @moduledoc false
+
+  use GenServer, restart: :temporary
+
+  @max_partial_request_count 12
+  @max_single_partial_request_retries 3
+
+  # in ms
+  @default_wait_period 1000
+
+  @spec start_link(%{url: String.t(), parent: GenServer.server(), name: String.t()}) ::
+          GenServer.on_start()
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @impl true
+  def init(opts) do
+    master_manifest_uri = URI.parse(opts.url)
+
+    state = %{
+      name: opts.name,
+      parent: opts.parent,
+      base_uri: Map.update!(master_manifest_uri, :path, &Path.dirname/1),
+      track_manifest_name: nil,
+      latest_partial: nil
+    }
+
+    {:ok, state, {:continue, {:get_master_manifest, master_manifest_uri}}}
+  end
+
+  @impl true
+  def handle_continue({:get_master_manifest, uri}, state) do
+    case request(uri, "master playlist", state.name, state.parent) do
+      {:ok, master_manifest} ->
+        send(self(), :get_new_partials)
+        track_manifest_name = get_track_manifest_name(master_manifest)
+
+        {:noreply, %{state | track_manifest_name: track_manifest_name}}
+
+      {:error, _response} ->
+        {:stop, :missing_master_manifest, state}
+    end
+  end
+
+  @impl true
+  def handle_info(:get_new_partials, state) do
+    latest_partial =
+      case request_track_manifest(state) do
+        {:ok, track_manifest} ->
+          track_manifest
+          |> get_new_partials_info(state.latest_partial)
+          |> Enum.reduce(state.latest_partial, fn partial, _acc ->
+            request_partial(partial, state)
+          end)
+
+        {:error, _response} ->
+          state.latest_partial
+      end
+
+    wait_for =
+      if is_nil(latest_partial),
+        do: @default_wait_period,
+        else: floor(latest_partial.duration * 1000)
+
+    Process.send_after(self(), :get_new_partials, wait_for)
```

Comment:
In the case of ll-hls we shouldn't wait between requests. Instead, we should pass parameters (HLS_msn, HLS_part) in HTTP requests. This way jellyfish will know that it has to block them and wait for the next partial